// @fivethree/core/esm2015/lib/overlay/overlay-content/overlay-content.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/overlay/overlay-content/overlay-content.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, HostBinding } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
export class FivOverlayContent {
    /**
     * @param {?} sanitizer
     */
    constructor(sanitizer) {
        this.sanitizer = sanitizer;
    }
    /**
     * @return {?}
     */
    get myStyle() {
        return this.priority
            ? this.sanitizer.bypassSecurityTrustStyle(`z-index: ${this.priority}`)
            : null;
    }
}
FivOverlayContent.decorators = [
    { type: Component, args: [{
                selector: 'fiv-overlay-content',
                template: "<ng-content #content></ng-content>",
                styles: [""]
            }] }
];
/** @nocollapse */
FivOverlayContent.ctorParameters = () => [
    { type: DomSanitizer }
];
FivOverlayContent.propDecorators = {
    myStyle: [{ type: HostBinding, args: ['style',] }]
};
if (false) {
    /** @type {?} */
    FivOverlayContent.prototype.priority;
    /**
     * @type {?}
     * @private
     */
    FivOverlayContent.prototype.sanitizer;
}
