// @fivethree/core/esm5/lib/if-platform/if-platform.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/if-platform/if-platform.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, ViewContainerRef, TemplateRef } from '@angular/core';
import { Platform } from '@ionic/angular';
var FivIfPlatform = /** @class */ (function () {
    function FivIfPlatform(platform, viewContainer, templateRef) {
        this.platform = platform;
        this.viewContainer = viewContainer;
        this.templateRef = templateRef;
    }
    Object.defineProperty(FivIfPlatform.prototype, "fivIf", {
        set: /**
         * @param {?} platforms
         * @return {?}
         */
        function (platforms) {
            var _this = this;
            /** @type {?} */
            var show = true;
            platforms.forEach((/**
             * @param {?} p
             * @return {?}
             */
            function (p) {
                if (!_this.platform.is(p)) {
                    show = false;
                }
            }));
            if (show) {
                this.viewContainer.createEmbeddedView(this.templateRef);
            }
            else {
                this.viewContainer.clear();
            }
        },
        enumerable: true,
        configurable: true
    });
    FivIfPlatform.decorators = [
        { type: Directive, args: [{
                    selector: '[fivIf]'
                },] }
    ];
    /** @nocollapse */
    FivIfPlatform.ctorParameters = function () { return [
        { type: Platform },
        { type: ViewContainerRef },
        { type: TemplateRef }
    ]; };
    FivIfPlatform.propDecorators = {
        fivIf: [{ type: Input }]
    };
    return FivIfPlatform;
}());
export { FivIfPlatform };
if (false) {
    /**
     * @type {?}
     * @private
     */
    FivIfPlatform.prototype.platform;
    /**
     * @type {?}
     * @private
     */
    FivIfPlatform.prototype.viewContainer;
    /**
     * @type {?}
     * @private
     */
    FivIfPlatform.prototype.templateRef;
}
