// @fivethree/core/esm5/lib/expandable/expandable.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/expandable/expandable.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Output, EventEmitter, Input, ChangeDetectorRef } from '@angular/core';
import { trigger, state, style, transition, animate, keyframes } from '@angular/animations';
var FivExpandable = /** @class */ (function () {
    function FivExpandable(change) {
        this.change = change;
        this.isOpen = false;
        this.fivWillOpen = new EventEmitter();
        this.fivDidOpen = new EventEmitter();
        this.fivWillClose = new EventEmitter();
        this.fivDidClose = new EventEmitter();
    }
    /**
     * @return {?}
     */
    FivExpandable.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @return {?}
     */
    FivExpandable.prototype.open = /**
     * @return {?}
     */
    function () {
        this.fivWillOpen.emit(this);
        this.isOpen = true;
        this.change.detectChanges();
    };
    /**
     * @return {?}
     */
    FivExpandable.prototype.toggle = /**
     * @return {?}
     */
    function () {
        if (this.isOpen === false) {
            this.open();
        }
        else {
            this.close();
        }
    };
    /**
     * @param {?=} param
     * @return {?}
     */
    FivExpandable.prototype.close = /**
     * @param {?=} param
     * @return {?}
     */
    function (param) {
        this.param = param;
        this.fivWillClose.emit(this);
        this.isOpen = false;
        this.change.detectChanges();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    FivExpandable.prototype.onAnimationEnd = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (event.fromState === 'closed') {
            this.fivDidOpen.emit(this);
        }
        else if (event.fromState === 'open') {
            this.fivDidClose.emit({ expandable: this, param: this.param });
            this.param = null;
        }
    };
    FivExpandable.decorators = [
        { type: Component, args: [{
                    selector: 'fiv-expandable',
                    template: "<div>\n  <ng-content select=\"[header]\"></ng-content>\n  <div [@listAnim]=\"{value: isOpen ? 'open' : 'closed', params: timingFunction ? {time: timingFunction} : {}}\" (@listAnim.done)=\"onAnimationEnd($event)\"\n    style=\"overflow: hidden\">\n    <ng-content select=\"[content]\"></ng-content>\n  </div>\n</div>",
                    animations: [
                        trigger('listAnim', [
                            state('open', style({ height: '*', opacity: 1 })),
                            state('closed', style({ height: '0', opacity: 0 })),
                            transition('closed => open', [
                                animate('{{time}}', keyframes([
                                    style({ height: '0', opacity: 0, offset: 0 }),
                                    style({ height: '*', opacity: 0.1, offset: 0.8 }),
                                    style({ height: '*', opacity: 1, offset: 1 })
                                ]))
                            ], { params: { time: '270ms ease-out' } }),
                            transition('open => closed', [
                                animate('{{time}}', keyframes([
                                    style({ height: '*', opacity: 1, offset: 0 }),
                                    style({ height: '*', opacity: 0.1, offset: 0.2 }),
                                    style({ height: '0', opacity: 0, offset: 1 })
                                ]))
                            ], { params: { time: '220ms ease-out' } })
                        ])
                    ],
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    FivExpandable.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    FivExpandable.propDecorators = {
        isOpen: [{ type: Input }],
        timingFunction: [{ type: Input }],
        fivWillOpen: [{ type: Output }],
        fivDidOpen: [{ type: Output }],
        fivWillClose: [{ type: Output }],
        fivDidClose: [{ type: Output }]
    };
    return FivExpandable;
}());
export { FivExpandable };
if (false) {
    /** @type {?} */
    FivExpandable.prototype.isOpen;
    /** @type {?} */
    FivExpandable.prototype.timingFunction;
    /** @type {?} */
    FivExpandable.prototype.fivWillOpen;
    /** @type {?} */
    FivExpandable.prototype.fivDidOpen;
    /** @type {?} */
    FivExpandable.prototype.fivWillClose;
    /** @type {?} */
    FivExpandable.prototype.fivDidClose;
    /** @type {?} */
    FivExpandable.prototype.param;
    /**
     * @type {?}
     * @private
     */
    FivExpandable.prototype.change;
}
