// @fivethree/core/esm2015/lib/gallery/keycodes.enum.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/gallery/keycodes.enum.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
const Key = {
    Backspace: 8,
    Tab: 9,
    Enter: 13,
    Shift: 16,
    Ctrl: 17,
    Alt: 18,
    PauseBreak: 19,
    CapsLock: 20,
    Escape: 27,
    Space: 32,
    PageUp: 33,
    PageDown: 34,
    End: 35,
    Home: 36,
    LeftArrow: 37,
    UpArrow: 38,
    RightArrow: 39,
    DownArrow: 40,
    Insert: 45,
    Delete: 46,
    Zero: 48,
    ClosedParen: 48,
    One: 49,
    ExclamationMark: 49,
    Two: 50,
    AtSign: 50,
    Three: 51,
    PoundSign: 51,
    Hash: 51,
    Four: 52,
    DollarSign: 52,
    Five: 53,
    PercentSign: 53,
    Six: 54,
    Caret: 54,
    Hat: 54,
    Seven: 55,
    Ampersand: 55,
    Eight: 56,
    Star: 56,
    Asterik: 56,
    Nine: 57,
    OpenParen: 57,
    A: 65,
    B: 66,
    C: 67,
    D: 68,
    E: 69,
    F: 70,
    G: 71,
    H: 72,
    I: 73,
    J: 74,
    K: 75,
    L: 76,
    M: 77,
    N: 78,
    O: 79,
    P: 80,
    Q: 81,
    R: 82,
    S: 83,
    T: 84,
    U: 85,
    V: 86,
    W: 87,
    X: 88,
    Y: 89,
    Z: 90,
    LeftWindowKey: 91,
    RightWindowKey: 92,
    SelectKey: 93,
    Numpad0: 96,
    Numpad1: 97,
    Numpad2: 98,
    Numpad3: 99,
    Numpad4: 100,
    Numpad5: 101,
    Numpad6: 102,
    Numpad7: 103,
    Numpad8: 104,
    Numpad9: 105,
    Multiply: 106,
    Add: 107,
    Subtract: 109,
    DecimalPoint: 110,
    Divide: 111,
    F1: 112,
    F2: 113,
    F3: 114,
    F4: 115,
    F5: 116,
    F6: 117,
    F7: 118,
    F8: 119,
    F9: 120,
    F10: 121,
    F11: 122,
    F12: 123,
    NumLock: 144,
    ScrollLock: 145,
    SemiColon: 186,
    Equals: 187,
    Comma: 188,
    Dash: 189,
    Period: 190,
    UnderScore: 189,
    PlusSign: 187,
    ForwardSlash: 191,
    Tilde: 192,
    GraveAccent: 192,
    OpenBracket: 219,
    ClosedBracket: 221,
    Quote: 222,
};
export { Key };
Key[Key.Backspace] = 'Backspace';
Key[Key.Tab] = 'Tab';
Key[Key.Enter] = 'Enter';
Key[Key.Shift] = 'Shift';
Key[Key.Ctrl] = 'Ctrl';
Key[Key.Alt] = 'Alt';
Key[Key.PauseBreak] = 'PauseBreak';
Key[Key.CapsLock] = 'CapsLock';
Key[Key.Escape] = 'Escape';
Key[Key.Space] = 'Space';
Key[Key.PageUp] = 'PageUp';
Key[Key.PageDown] = 'PageDown';
Key[Key.End] = 'End';
Key[Key.Home] = 'Home';
Key[Key.LeftArrow] = 'LeftArrow';
Key[Key.UpArrow] = 'UpArrow';
Key[Key.RightArrow] = 'RightArrow';
Key[Key.DownArrow] = 'DownArrow';
Key[Key.Insert] = 'Insert';
Key[Key.Delete] = 'Delete';
Key[Key.Zero] = 'Zero';
Key[Key.ClosedParen] = 'ClosedParen';
Key[Key.One] = 'One';
Key[Key.ExclamationMark] = 'ExclamationMark';
Key[Key.Two] = 'Two';
Key[Key.AtSign] = 'AtSign';
Key[Key.Three] = 'Three';
Key[Key.PoundSign] = 'PoundSign';
Key[Key.Hash] = 'Hash';
Key[Key.Four] = 'Four';
Key[Key.DollarSign] = 'DollarSign';
Key[Key.Five] = 'Five';
Key[Key.PercentSign] = 'PercentSign';
Key[Key.Six] = 'Six';
Key[Key.Caret] = 'Caret';
Key[Key.Hat] = 'Hat';
Key[Key.Seven] = 'Seven';
Key[Key.Ampersand] = 'Ampersand';
Key[Key.Eight] = 'Eight';
Key[Key.Star] = 'Star';
Key[Key.Asterik] = 'Asterik';
Key[Key.Nine] = 'Nine';
Key[Key.OpenParen] = 'OpenParen';
Key[Key.A] = 'A';
Key[Key.B] = 'B';
Key[Key.C] = 'C';
Key[Key.D] = 'D';
Key[Key.E] = 'E';
Key[Key.F] = 'F';
Key[Key.G] = 'G';
Key[Key.H] = 'H';
Key[Key.I] = 'I';
Key[Key.J] = 'J';
Key[Key.K] = 'K';
Key[Key.L] = 'L';
Key[Key.M] = 'M';
Key[Key.N] = 'N';
Key[Key.O] = 'O';
Key[Key.P] = 'P';
Key[Key.Q] = 'Q';
Key[Key.R] = 'R';
Key[Key.S] = 'S';
Key[Key.T] = 'T';
Key[Key.U] = 'U';
Key[Key.V] = 'V';
Key[Key.W] = 'W';
Key[Key.X] = 'X';
Key[Key.Y] = 'Y';
Key[Key.Z] = 'Z';
Key[Key.LeftWindowKey] = 'LeftWindowKey';
Key[Key.RightWindowKey] = 'RightWindowKey';
Key[Key.SelectKey] = 'SelectKey';
Key[Key.Numpad0] = 'Numpad0';
Key[Key.Numpad1] = 'Numpad1';
Key[Key.Numpad2] = 'Numpad2';
Key[Key.Numpad3] = 'Numpad3';
Key[Key.Numpad4] = 'Numpad4';
Key[Key.Numpad5] = 'Numpad5';
Key[Key.Numpad6] = 'Numpad6';
Key[Key.Numpad7] = 'Numpad7';
Key[Key.Numpad8] = 'Numpad8';
Key[Key.Numpad9] = 'Numpad9';
Key[Key.Multiply] = 'Multiply';
Key[Key.Add] = 'Add';
Key[Key.Subtract] = 'Subtract';
Key[Key.DecimalPoint] = 'DecimalPoint';
Key[Key.Divide] = 'Divide';
Key[Key.F1] = 'F1';
Key[Key.F2] = 'F2';
Key[Key.F3] = 'F3';
Key[Key.F4] = 'F4';
Key[Key.F5] = 'F5';
Key[Key.F6] = 'F6';
Key[Key.F7] = 'F7';
Key[Key.F8] = 'F8';
Key[Key.F9] = 'F9';
Key[Key.F10] = 'F10';
Key[Key.F11] = 'F11';
Key[Key.F12] = 'F12';
Key[Key.NumLock] = 'NumLock';
Key[Key.ScrollLock] = 'ScrollLock';
Key[Key.SemiColon] = 'SemiColon';
Key[Key.Equals] = 'Equals';
Key[Key.Comma] = 'Comma';
Key[Key.Dash] = 'Dash';
Key[Key.Period] = 'Period';
Key[Key.UnderScore] = 'UnderScore';
Key[Key.PlusSign] = 'PlusSign';
Key[Key.ForwardSlash] = 'ForwardSlash';
Key[Key.Tilde] = 'Tilde';
Key[Key.GraveAccent] = 'GraveAccent';
Key[Key.OpenBracket] = 'OpenBracket';
Key[Key.ClosedBracket] = 'ClosedBracket';
Key[Key.Quote] = 'Quote';
