// @fivethree/core/esm2015/lib/if-platform/if-platform.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/if-platform/if-platform.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, ViewContainerRef, TemplateRef } from '@angular/core';
import { Platform } from '@ionic/angular';
export class FivIfPlatform {
    /**
     * @param {?} platform
     * @param {?} viewContainer
     * @param {?} templateRef
     */
    constructor(platform, viewContainer, templateRef) {
        this.platform = platform;
        this.viewContainer = viewContainer;
        this.templateRef = templateRef;
    }
    /**
     * @param {?} platforms
     * @return {?}
     */
    set fivIf(platforms) {
        /** @type {?} */
        let show = true;
        platforms.forEach((/**
         * @param {?} p
         * @return {?}
         */
        p => {
            if (!this.platform.is(p)) {
                show = false;
            }
        }));
        if (show) {
            this.viewContainer.createEmbeddedView(this.templateRef);
        }
        else {
            this.viewContainer.clear();
        }
    }
}
FivIfPlatform.decorators = [
    { type: Directive, args: [{
                selector: '[fivIf]'
            },] }
];
/** @nocollapse */
FivIfPlatform.ctorParameters = () => [
    { type: Platform },
    { type: ViewContainerRef },
    { type: TemplateRef }
];
FivIfPlatform.propDecorators = {
    fivIf: [{ type: Input }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    FivIfPlatform.prototype.platform;
    /**
     * @type {?}
     * @private
     */
    FivIfPlatform.prototype.viewContainer;
    /**
     * @type {?}
     * @private
     */
    FivIfPlatform.prototype.templateRef;
}
