// @fivethree/core/esm5/lib/network-status/network-status.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/network-status/network-status.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter } from '@angular/core';
import { trigger, transition, style, animate } from '@angular/animations';
var FivNetworkStatus = /** @class */ (function () {
    function FivNetworkStatus() {
        this.visible = true;
        this.status = 'online';
        this.onClick = new EventEmitter();
    }
    /**
     * @return {?}
     */
    FivNetworkStatus.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @return {?}
     */
    FivNetworkStatus.prototype.show = /**
     * @return {?}
     */
    function () {
        this.visible = true;
    };
    /**
     * @return {?}
     */
    FivNetworkStatus.prototype.hide = /**
     * @return {?}
     */
    function () {
        this.visible = false;
    };
    /**
     * @param {?} status
     * @return {?}
     */
    FivNetworkStatus.prototype.setStatus = /**
     * @param {?} status
     * @return {?}
     */
    function (status) {
        this.status = status;
    };
    /**
     * @return {?}
     */
    FivNetworkStatus.prototype.FivNetworkStatusClicked = /**
     * @return {?}
     */
    function () {
        this.onClick.emit(this);
    };
    FivNetworkStatus.decorators = [
        { type: Component, args: [{
                    selector: 'fiv-network-status',
                    template: "<ion-chip [@status] (click)=\"FivNetworkStatusClicked()\" *ngIf=\"visible\">\n  <fiv-icon [badge]=\"-1\" [fivCenter] [off]=\"status === 'offline'\" name=\"cloud\"></fiv-icon>\n</ion-chip>",
                    animations: [
                        trigger('status', [
                            transition('void => *', [
                                style({ opacity: 0 }),
                                animate('200ms ease-out', style({ opacity: 1 }))
                            ]),
                            transition('* => void', [
                                style({ opacity: 1 }),
                                animate('550ms 200ms ease-in', style({ opacity: 0 }))
                            ])
                        ])
                    ],
                    styles: [":host{position:absolute;top:66px;right:8px}ion-chip{width:56px}fiv-icon{--fiv-icon-color:var(--ion-color-dark)}"]
                }] }
    ];
    /** @nocollapse */
    FivNetworkStatus.ctorParameters = function () { return []; };
    return FivNetworkStatus;
}());
export { FivNetworkStatus };
if (false) {
    /** @type {?} */
    FivNetworkStatus.prototype.visible;
    /** @type {?} */
    FivNetworkStatus.prototype.status;
    /** @type {?} */
    FivNetworkStatus.prototype.onClick;
}
