// @fivethree/core/esm5/lib/collapsable-menu/collapsable-menu-button/collapsable-menu-button.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/collapsable-menu/collapsable-menu-button/collapsable-menu-button.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { FivCollapsableMenu } from '../collapsable-menu.directive';
var FivCollapsableMenuButton = /** @class */ (function () {
    function FivCollapsableMenuButton() {
        this.icon = 'menu';
    }
    /**
     * @return {?}
     */
    FivCollapsableMenuButton.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @return {?}
     */
    FivCollapsableMenuButton.prototype.onClick = /**
     * @return {?}
     */
    function () {
        this.menu.toggleMenu();
    };
    FivCollapsableMenuButton.decorators = [
        { type: Component, args: [{
                    selector: 'fiv-collapsable-menu-button',
                    template: "<ion-button fivCenter (click)=\"onClick()\">\n  <fiv-icon slot=\"icon-only\" [name]=\"icon\" [color]=\"color\"></fiv-icon>\n</ion-button>",
                    styles: [":host{display:block;padding-top:.2em}"]
                }] }
    ];
    /** @nocollapse */
    FivCollapsableMenuButton.ctorParameters = function () { return []; };
    FivCollapsableMenuButton.propDecorators = {
        menu: [{ type: Input }],
        icon: [{ type: Input }],
        color: [{ type: Input }]
    };
    return FivCollapsableMenuButton;
}());
export { FivCollapsableMenuButton };
if (false) {
    /** @type {?} */
    FivCollapsableMenuButton.prototype.menu;
    /** @type {?} */
    FivCollapsableMenuButton.prototype.icon;
    /** @type {?} */
    FivCollapsableMenuButton.prototype.color;
}
