// @fivethree/core/esm2015/lib/stepper/step-content/step-content.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/stepper/step-content/step-content.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild, Input, Output, EventEmitter, ChangeDetectorRef } from '@angular/core';
import { FivStepHeader } from '../step-header/step-header.component';
import { FivExpandable } from '../../expandable/expandable.component';
export class FivStepContent {
    /**
     * @param {?} change
     */
    constructor(change) {
        this.change = change;
        this.isLast = false;
        this.isOpen = false;
        this.title = '';
        this.subtitle = '';
        this.fivDidOpen = new EventEmitter();
        this.fivDidClose = new EventEmitter();
        this.fivClick = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @return {?}
     */
    open() {
        this.isOpen = true;
        this.change.detectChanges();
    }
    /**
     * @return {?}
     */
    close() {
        this.isOpen = false;
        this.change.detectChanges();
    }
    /**
     * @return {?}
     */
    completeAndClose() {
        this.close();
        this.complete();
    }
    /**
     * @return {?}
     */
    complete() {
        this.header.complete();
    }
    /**
     * @return {?}
     */
    reset() {
        this.header.reset();
    }
    /**
     * @return {?}
     */
    afterClose() {
        this.fivDidClose.emit(this);
    }
    /**
     * @return {?}
     */
    afterOpen() {
        this.fivDidOpen.emit(this);
    }
    /**
     * @return {?}
     */
    click() {
        this.fivClick.emit(this);
    }
}
FivStepContent.decorators = [
    { type: Component, args: [{
                selector: 'fiv-step-content',
                template: "<div *ngIf=\"!isLast\" class=\"timeline\"></div>\n<fiv-expandable [isOpen]=\"isOpen\" (fivDidClose)=\"afterClose()\" (fivDidOpen)=\"afterOpen()\" #self>\n  <div header class=\"ion-activatable\" (click)=\"click()\">\n    <fiv-step-header #header [index]=\"index\" [icon]=\"icon\">\n      <div class=\"header-title\">{{title}}</div>\n      <div class=\"header-subtitle\">{{subtitle}}</div>\n\n      <ion-ripple-effect type=\"bounded\"></ion-ripple-effect>\n    </fiv-step-header>\n  </div>\n  <ng-container content>\n    <div class=\"step-content\">\n      <ng-content></ng-content>\n    </div>\n  </ng-container>\n</fiv-expandable>",
                styles: [":host{display:block;width:100%;position:relative}.step-content{padding-left:44px;padding-bottom:24px}.timeline{width:1px;height:calc(100% - 32px);position:absolute;background:var(--fiv-color-timeline,var(--ion-color-medium));left:27.5px;top:36px}.ion-activatable{position:relative;overflow:hidden}.header-title{color:var(--fiv-color-title,var(--ion-color-dark))}.header-subtitle{color:var(--fiv-color-subtitle,var(--ion-color-medium));font-size:.8em}"]
            }] }
];
/** @nocollapse */
FivStepContent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
FivStepContent.propDecorators = {
    index: [{ type: Input }],
    icon: [{ type: Input }],
    isLast: [{ type: Input }],
    isOpen: [{ type: Input }],
    title: [{ type: Input }],
    subtitle: [{ type: Input }],
    fivDidOpen: [{ type: Output }],
    fivDidClose: [{ type: Output }],
    fivClick: [{ type: Output }],
    step: [{ type: ViewChild, args: ['self', { static: false },] }],
    header: [{ type: ViewChild, args: ['header', { static: false },] }]
};
if (false) {
    /** @type {?} */
    FivStepContent.prototype.index;
    /** @type {?} */
    FivStepContent.prototype.icon;
    /** @type {?} */
    FivStepContent.prototype.isLast;
    /** @type {?} */
    FivStepContent.prototype.isOpen;
    /** @type {?} */
    FivStepContent.prototype.title;
    /** @type {?} */
    FivStepContent.prototype.subtitle;
    /** @type {?} */
    FivStepContent.prototype.fivDidOpen;
    /** @type {?} */
    FivStepContent.prototype.fivDidClose;
    /** @type {?} */
    FivStepContent.prototype.fivClick;
    /** @type {?} */
    FivStepContent.prototype.step;
    /** @type {?} */
    FivStepContent.prototype.header;
    /**
     * @type {?}
     * @private
     */
    FivStepContent.prototype.change;
}
