// @fivethree/core/esm5/lib/gallery/gallery-image/gallery-image.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/gallery/gallery-image/gallery-image.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ElementRef, ViewChild, Output, EventEmitter } from '@angular/core';
var FivGalleryImage = /** @class */ (function () {
    function FivGalleryImage() {
        this.click = new EventEmitter();
    }
    /**
     * @return {?}
     */
    FivGalleryImage.prototype.open = /**
     * @return {?}
     */
    function () {
        this.click.emit(this);
    };
    /**
     * @return {?}
     */
    FivGalleryImage.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    FivGalleryImage.decorators = [
        { type: Component, args: [{
                    selector: 'fiv-gallery-image',
                    template: "<ng-container *ngIf=\"!src\">\n  <ng-content></ng-content>\n</ng-container>\n<div class=\"ion-activatable\" (click)=\"open()\">\n  <img #thumbnail class=\"thumbnail\" [src]=\"src\">\n  <ion-ripple-effect type=\"bounded\"></ion-ripple-effect>\n</div>",
                    styles: [":host{display:block;--max-height:100%;--border-radius:0px;min-height:var(--max-height);--background:var(--ion-color-light);background:var(--background)}.thumbnail{-o-object-position:center;object-position:center;-o-object-fit:cover;object-fit:cover;display:block;height:var(--height,auto);max-height:var(--max-height);width:var(--width,100%);border-radius:var(--border-radius)}.ion-activatable{position:relative;--ripple-color:rgba(0, 0, 0, 0.5)}"]
                }] }
    ];
    /** @nocollapse */
    FivGalleryImage.ctorParameters = function () { return []; };
    FivGalleryImage.propDecorators = {
        src: [{ type: Input }],
        thumbnail: [{ type: ViewChild, args: ['thumbnail', { static: true },] }],
        click: [{ type: Output }]
    };
    return FivGalleryImage;
}());
export { FivGalleryImage };
if (false) {
    /** @type {?} */
    FivGalleryImage.prototype.src;
    /** @type {?} */
    FivGalleryImage.prototype.originalSrc;
    /** @type {?} */
    FivGalleryImage.prototype.index;
    /** @type {?} */
    FivGalleryImage.prototype.thumbnail;
    /** @type {?} */
    FivGalleryImage.prototype.click;
}
