// @fivethree/core/esm5/lib/lazy-image/lazy-image.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/lazy-image/lazy-image.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { FivGalleryImage } from '../gallery/gallery-image/gallery-image.component';
import { Directive, Input, EventEmitter, Output, Host, Self, Optional, ElementRef } from '@angular/core';
var FivLazyImage = /** @class */ (function () {
    function FivLazyImage(fivImage, image) {
        this.fivImage = fivImage;
        this.image = image;
        this.virtual = false;
        this.inViewport = false;
        this.willShow = new EventEmitter();
        this.willHide = new EventEmitter();
    }
    /**
     * @return {?}
     */
    FivLazyImage.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var io = new IntersectionObserver((/**
         * @param {?} entries
         * @param {?} observer
         * @return {?}
         */
        function (entries, observer) {
            entries.forEach((/**
             * @param {?} entry
             * @return {?}
             */
            function (entry) {
                if (entry.isIntersecting) {
                    if (!_this.inViewport) {
                        _this.handleIntersection();
                        if (!_this.virtual) {
                            io.disconnect();
                        }
                    }
                }
                else {
                    if (_this.inViewport) {
                        _this.inViewport = false;
                        if (_this.fivImage) {
                            _this.willHide.emit(_this.fivImage);
                        }
                        else if (_this.image) {
                            _this.willHide.emit(_this.image.nativeElement);
                        }
                    }
                }
            }));
        }));
        io['POLL_INTERVAL'] = 100;
        if (this.fivImage) {
            io.observe(this.fivImage.thumbnail.nativeElement);
        }
        else if (this.image) {
            io.observe(this.image.nativeElement);
        }
    };
    /**
     * @return {?}
     */
    FivLazyImage.prototype.handleIntersection = /**
     * @return {?}
     */
    function () {
        this.inViewport = true;
        if (this.fivLazyImage) {
            if (this.fivImage) {
                this.fivImage.src = this.fivLazyImage;
                this.willShow.emit(this.fivImage);
            }
            else if (this.image) {
                this.image.nativeElement.src = this.fivLazyImage;
                this.willShow.emit(this.image.nativeElement);
            }
        }
        else {
            if (this.fivImage) {
                this.willShow.emit(this.fivImage);
            }
            else if (this.image) {
                this.willShow.emit(this.image.nativeElement);
            }
        }
    };
    FivLazyImage.decorators = [
        { type: Directive, args: [{
                    selector: '[fivLazyImage]'
                },] }
    ];
    /** @nocollapse */
    FivLazyImage.ctorParameters = function () { return [
        { type: FivGalleryImage, decorators: [{ type: Host }, { type: Self }, { type: Optional }] },
        { type: ElementRef, decorators: [{ type: Host }, { type: Self }, { type: Optional }] }
    ]; };
    FivLazyImage.propDecorators = {
        fivLazyImage: [{ type: Input }],
        virtual: [{ type: Input }],
        willShow: [{ type: Output }],
        willHide: [{ type: Output }]
    };
    return FivLazyImage;
}());
export { FivLazyImage };
if (false) {
    /** @type {?} */
    FivLazyImage.prototype.fivLazyImage;
    /** @type {?} */
    FivLazyImage.prototype.virtual;
    /** @type {?} */
    FivLazyImage.prototype.inViewport;
    /** @type {?} */
    FivLazyImage.prototype.willShow;
    /** @type {?} */
    FivLazyImage.prototype.willHide;
    /** @type {?} */
    FivLazyImage.prototype.fivImage;
    /** @type {?} */
    FivLazyImage.prototype.image;
}
