// @fivethree/core/esm2015/lib/app-bar/app-bar-tab/app-bar-tab.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/app-bar/app-bar-tab/app-bar-tab.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Host, HostBinding, ElementRef } from '@angular/core';
import { Router } from '@angular/router';
import { IonTabs } from '@ionic/angular';
import { FivAppBar } from '../app-bar.component';
export class FivAppBarTab {
    /**
     * @param {?} appBar
     * @param {?} ionTabs
     * @param {?} router
     * @param {?} el
     */
    constructor(appBar, ionTabs, router, el) {
        this.appBar = appBar;
        this.ionTabs = ionTabs;
        this.router = router;
        this.el = el;
        this.type = 'unbounded';
    }
    /**
     * @return {?}
     */
    get active() {
        return this.router.url.endsWith(this.href);
    }
    /**
     * @return {?}
     */
    get classes() {
        return `label-${this.appBar.titleMode}`;
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @return {?}
     */
    onClick() {
        this.ionTabs.select(this.tab);
    }
}
FivAppBarTab.decorators = [
    { type: Component, args: [{
                selector: 'fiv-app-bar-tab',
                template: "<ng-template>\n  <ng-content></ng-content>\n</ng-template>\n\n<div (click)=\"onClick()\" class=\"tab ion-activatable\" [ngClass]=\"{'active': active}\">\n  <ng-content select=\"fiv-icon\"></ng-content>\n  <ng-content *ngIf=\"appBar.titleMode === 'show' || (active && appBar.titleMode === 'active-only')\" select=\"ion-label\">\n  </ng-content>\n  <ion-ripple-effect [type]=\"type\"></ion-ripple-effect>\n</div>",
                styles: [":host{width:50%;height:100%;display:inline-block;position:relative}:host .tab{width:100%;height:100%;display:flex;position:relative;flex-direction:column;align-items:center;justify-content:center;overflow:visible}:host .tab.active{--fiv-icon-color:var(--fiv-icon-activ-color, var(--ion-color-primary))}:host .tab.active ::ng-deep ion-label{color:var(--ion-color-primary)}:host(.label-hide) ::ng-deep ion-label{display:none}:host(.label-active-only) .tab:not(.active) ::ng-deep ion-label{display:none}"]
            }] }
];
/** @nocollapse */
FivAppBarTab.ctorParameters = () => [
    { type: FivAppBar, decorators: [{ type: Host }] },
    { type: IonTabs, decorators: [{ type: Host }] },
    { type: Router },
    { type: ElementRef }
];
FivAppBarTab.propDecorators = {
    tab: [{ type: Input }],
    href: [{ type: Input }],
    type: [{ type: Input }],
    classes: [{ type: HostBinding, args: ['class',] }]
};
if (false) {
    /** @type {?} */
    FivAppBarTab.prototype.tab;
    /** @type {?} */
    FivAppBarTab.prototype.href;
    /** @type {?} */
    FivAppBarTab.prototype.type;
    /** @type {?} */
    FivAppBarTab.prototype.appBar;
    /** @type {?} */
    FivAppBarTab.prototype.ionTabs;
    /** @type {?} */
    FivAppBarTab.prototype.router;
    /** @type {?} */
    FivAppBarTab.prototype.el;
}
