// @fivethree/core/esm5/lib/overlay/overlay.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/overlay/overlay.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FivOverlay } from './overlay.component';
import { FivOverlayContent } from './overlay-content/overlay-content.component';
import { FivOverlayService } from './overlay.service';
var FivOverlayModule = /** @class */ (function () {
    function FivOverlayModule() {
    }
    FivOverlayModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [FivOverlay, FivOverlayContent],
                    imports: [CommonModule],
                    exports: [FivOverlay, FivOverlayContent],
                    entryComponents: [FivOverlayContent],
                    providers: [FivOverlayService]
                },] }
    ];
    return FivOverlayModule;
}());
export { FivOverlayModule };
