// @fivethree/core/esm5/lib/gallery/gallery-toolbar-content/gallery-toolbar-content.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/gallery/gallery-toolbar-content/gallery-toolbar-content.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { FivGallery } from './../gallery.component';
import { Component, Host } from '@angular/core';
var FivGalleryToolbarContent = /** @class */ (function () {
    function FivGalleryToolbarContent(gallery) {
        this.gallery = gallery;
    }
    /**
     * @return {?}
     */
    FivGalleryToolbarContent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    FivGalleryToolbarContent.decorators = [
        { type: Component, args: [{
                    selector: 'fiv-gallery-toolbar-content',
                    template: "<ion-toolbar color=\"transparent\">\n  <ion-buttons slot=\"start\">\n    <ion-button (click)=\"gallery.close()\">\n      <ion-icon slot=\"icon-only\" name=\"close\"></ion-icon>\n    </ion-button>\n  </ion-buttons>\n  <ion-buttons slot=\"end\">\n    <ion-button (click)=\"gallery.fullscreen()\">\n      <fiv-icon slot=\"icon-only\" [name]=\"gallery.inFullscreen ? 'contract': 'expand'\"></fiv-icon>\n    </ion-button>\n    <ion-button (click)=\"gallery.zoom()\">\n      <fiv-icon slot=\"icon-only\" [name]=\"gallery.zoomedIn ? 'remove': 'add'\"></fiv-icon>\n    </ion-button>\n  </ion-buttons>\n</ion-toolbar>\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    FivGalleryToolbarContent.ctorParameters = function () { return [
        { type: FivGallery, decorators: [{ type: Host }] }
    ]; };
    return FivGalleryToolbarContent;
}());
export { FivGalleryToolbarContent };
if (false) {
    /** @type {?} */
    FivGalleryToolbarContent.prototype.gallery;
}
