// @fivethree/core/esm5/lib/back-button/back-button.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/back-button/back-button.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { FivRoutingStateService } from './routing-state.service';
import { Component, Input } from '@angular/core';
var FivBackButton = /** @class */ (function () {
    function FivBackButton(routingState) {
        this.routingState = routingState;
        this.icon = 'arrow-back';
        this.defaultHref = '/';
    }
    /**
     * @return {?}
     */
    FivBackButton.prototype.buttonClick = /**
     * @return {?}
     */
    function () {
        this.routingState.goBack(this.defaultHref);
    };
    FivBackButton.decorators = [
        { type: Component, args: [{
                    selector: 'fiv-back-button',
                    template: "<ion-button (click)=\"buttonClick()\">\n  <fiv-icon slot=\"icon-only\" [color]=\"color\" [name]=\"icon\"></fiv-icon>\n</ion-button>",
                    styles: [":host{display:block;padding-top:.2em}"]
                }] }
    ];
    /** @nocollapse */
    FivBackButton.ctorParameters = function () { return [
        { type: FivRoutingStateService }
    ]; };
    FivBackButton.propDecorators = {
        icon: [{ type: Input }],
        color: [{ type: Input }],
        defaultHref: [{ type: Input }]
    };
    return FivBackButton;
}());
export { FivBackButton };
if (false) {
    /** @type {?} */
    FivBackButton.prototype.icon;
    /** @type {?} */
    FivBackButton.prototype.color;
    /** @type {?} */
    FivBackButton.prototype.defaultHref;
    /** @type {?} */
    FivBackButton.prototype.routingState;
}
