// @fivethree/core/esm5/lib/dialog/dialog.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/dialog/dialog.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { FivPullModule } from './../pull/pull.module';
import { FivDialogService } from './dialog.service';
import { FivOverlayModule } from './../overlay/overlay.module';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FivDialog } from './dialog.component';
import { IonicModule } from '@ionic/angular';
import { FivLoadingProgressBarModule } from '../loading-progress-bar/loading-progress-bar.module';
var FivDialogModule = /** @class */ (function () {
    function FivDialogModule() {
    }
    FivDialogModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [FivDialog],
                    imports: [
                        CommonModule,
                        FivLoadingProgressBarModule,
                        FivOverlayModule,
                        IonicModule,
                        FivPullModule
                    ],
                    exports: [FivDialog],
                    entryComponents: [FivDialog],
                    providers: [FivDialogService]
                },] }
    ];
    return FivDialogModule;
}());
export { FivDialogModule };
