// @fivethree/core/esm5/lib/dialog/dialog.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/dialog/dialog.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { FivOverlayService } from '../overlay/overlay.service';
import { FivDialog } from '../dialog/dialog.component';
import * as i0 from "@angular/core";
import * as i1 from "../overlay/overlay.service";
/**
 * @record
 */
export function DialogOptions() { }
if (false) {
    /** @type {?} */
    DialogOptions.prototype.verticalAlign;
    /** @type {?} */
    DialogOptions.prototype.horizontalAlign;
    /** @type {?} */
    DialogOptions.prototype.shape;
    /** @type {?} */
    DialogOptions.prototype.backdrop;
    /** @type {?} */
    DialogOptions.prototype.duration;
}
var FivDialogService = /** @class */ (function () {
    function FivDialogService(overlay) {
        this.overlay = overlay;
    }
    /**
     * @param {?} viewContainerRef
     * @param {?} content
     * @param {?=} options
     * @return {?}
     */
    FivDialogService.prototype.openDialog = /**
     * @param {?} viewContainerRef
     * @param {?} content
     * @param {?=} options
     * @return {?}
     */
    function (viewContainerRef, content, options) {
        /** @type {?} */
        var c = this.overlay.createOverlay(viewContainerRef, FivDialog, content);
        c.instance.verticalAlign = options ? options.verticalAlign : 'top';
        c.instance.verticalAlign = options ? options.verticalAlign : 'center';
        c.instance.backdrop = options ? options.backdrop : true;
        c.instance.duration = options ? options.duration : 0;
        c.instance.shape = options ? options.shape : 'card';
        c.instance.open();
        return c;
    };
    FivDialogService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    FivDialogService.ctorParameters = function () { return [
        { type: FivOverlayService }
    ]; };
    /** @nocollapse */ FivDialogService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function FivDialogService_Factory() { return new FivDialogService(i0.ɵɵinject(i1.FivOverlayService)); }, token: FivDialogService, providedIn: "root" });
    return FivDialogService;
}());
export { FivDialogService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    FivDialogService.prototype.overlay;
}
