// @fivethree/core/esm5/lib/stepper/step-content/step-content.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/stepper/step-content/step-content.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild, Input, Output, EventEmitter, ChangeDetectorRef } from '@angular/core';
import { FivStepHeader } from '../step-header/step-header.component';
import { FivExpandable } from '../../expandable/expandable.component';
var FivStepContent = /** @class */ (function () {
    function FivStepContent(change) {
        this.change = change;
        this.isLast = false;
        this.isOpen = false;
        this.title = '';
        this.subtitle = '';
        this.fivDidOpen = new EventEmitter();
        this.fivDidClose = new EventEmitter();
        this.fivClick = new EventEmitter();
    }
    /**
     * @return {?}
     */
    FivStepContent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @return {?}
     */
    FivStepContent.prototype.open = /**
     * @return {?}
     */
    function () {
        this.isOpen = true;
        this.change.detectChanges();
    };
    /**
     * @return {?}
     */
    FivStepContent.prototype.close = /**
     * @return {?}
     */
    function () {
        this.isOpen = false;
        this.change.detectChanges();
    };
    /**
     * @return {?}
     */
    FivStepContent.prototype.completeAndClose = /**
     * @return {?}
     */
    function () {
        this.close();
        this.complete();
    };
    /**
     * @return {?}
     */
    FivStepContent.prototype.complete = /**
     * @return {?}
     */
    function () {
        this.header.complete();
    };
    /**
     * @return {?}
     */
    FivStepContent.prototype.reset = /**
     * @return {?}
     */
    function () {
        this.header.reset();
    };
    /**
     * @return {?}
     */
    FivStepContent.prototype.afterClose = /**
     * @return {?}
     */
    function () {
        this.fivDidClose.emit(this);
    };
    /**
     * @return {?}
     */
    FivStepContent.prototype.afterOpen = /**
     * @return {?}
     */
    function () {
        this.fivDidOpen.emit(this);
    };
    /**
     * @return {?}
     */
    FivStepContent.prototype.click = /**
     * @return {?}
     */
    function () {
        this.fivClick.emit(this);
    };
    FivStepContent.decorators = [
        { type: Component, args: [{
                    selector: 'fiv-step-content',
                    template: "<div *ngIf=\"!isLast\" class=\"timeline\"></div>\n<fiv-expandable [isOpen]=\"isOpen\" (fivDidClose)=\"afterClose()\" (fivDidOpen)=\"afterOpen()\" #self>\n  <div header class=\"ion-activatable\" (click)=\"click()\">\n    <fiv-step-header #header [index]=\"index\" [icon]=\"icon\">\n      <div class=\"header-title\">{{title}}</div>\n      <div class=\"header-subtitle\">{{subtitle}}</div>\n\n      <ion-ripple-effect type=\"bounded\"></ion-ripple-effect>\n    </fiv-step-header>\n  </div>\n  <ng-container content>\n    <div class=\"step-content\">\n      <ng-content></ng-content>\n    </div>\n  </ng-container>\n</fiv-expandable>",
                    styles: [":host{display:block;width:100%;position:relative}.step-content{padding-left:44px;padding-bottom:24px}.timeline{width:1px;height:calc(100% - 32px);position:absolute;background:var(--fiv-color-timeline,var(--ion-color-medium));left:27.5px;top:36px}.ion-activatable{position:relative;overflow:hidden}.header-title{color:var(--fiv-color-title,var(--ion-color-dark))}.header-subtitle{color:var(--fiv-color-subtitle,var(--ion-color-medium));font-size:.8em}"]
                }] }
    ];
    /** @nocollapse */
    FivStepContent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    FivStepContent.propDecorators = {
        index: [{ type: Input }],
        icon: [{ type: Input }],
        isLast: [{ type: Input }],
        isOpen: [{ type: Input }],
        title: [{ type: Input }],
        subtitle: [{ type: Input }],
        fivDidOpen: [{ type: Output }],
        fivDidClose: [{ type: Output }],
        fivClick: [{ type: Output }],
        step: [{ type: ViewChild, args: ['self', { static: false },] }],
        header: [{ type: ViewChild, args: ['header', { static: false },] }]
    };
    return FivStepContent;
}());
export { FivStepContent };
if (false) {
    /** @type {?} */
    FivStepContent.prototype.index;
    /** @type {?} */
    FivStepContent.prototype.icon;
    /** @type {?} */
    FivStepContent.prototype.isLast;
    /** @type {?} */
    FivStepContent.prototype.isOpen;
    /** @type {?} */
    FivStepContent.prototype.title;
    /** @type {?} */
    FivStepContent.prototype.subtitle;
    /** @type {?} */
    FivStepContent.prototype.fivDidOpen;
    /** @type {?} */
    FivStepContent.prototype.fivDidClose;
    /** @type {?} */
    FivStepContent.prototype.fivClick;
    /** @type {?} */
    FivStepContent.prototype.step;
    /** @type {?} */
    FivStepContent.prototype.header;
    /**
     * @type {?}
     * @private
     */
    FivStepContent.prototype.change;
}
