// @fivethree/core/esm2015/lib/bottom-sheet/bottom-sheet-content/bottom-sheet-content.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bottom-sheet/bottom-sheet-content/bottom-sheet-content.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { IonContent } from '@ionic/angular';
import { DomSanitizer } from '@angular/platform-browser';
import { Component, Input, HostBinding, EventEmitter, Output, ViewChild } from '@angular/core';
import { DrawerState } from '../drawer-state';
export class FivBottomSheetContent {
    /**
     * @param {?} sanitizer
     */
    constructor(sanitizer) {
        this.sanitizer = sanitizer;
        this.rounded = true;
        this.handle = true;
        this.float = true;
        this.fivClick = new EventEmitter();
        this.states = DrawerState;
        this.shouldBounce = true;
        this.panning = false;
    }
    /**
     * @return {?}
     */
    get isRounded() {
        switch (this.currentState) {
            case DrawerState.Top:
                return false;
            case DrawerState.Docked:
                return this.rounded;
            case DrawerState.Bottom:
                return this.rounded && !this.float;
        }
        return true;
    }
    /**
     * @return {?}
     */
    get styles() {
        if (!(this.float && this.currentState === DrawerState.Bottom)) {
            return this.sanitizer
                .bypassSecurityTrustStyle(`box-shadow: 0 4px 16px rgba(0, 0, 0, .12);
                                                      border-top: 1px solid var(--ion-color-light);`);
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.content.scrollEvents = true;
    }
    /**
     * @param {?} state
     * @return {?}
     */
    updateState(state) {
        this.currentState = state;
    }
    /**
     * @return {?}
     */
    click() {
        this.fivClick.emit();
    }
}
FivBottomSheetContent.decorators = [
    { type: Component, args: [{
                selector: 'fiv-bottom-sheet-content',
                template: "<div class=\"bsc-wrapper\">\n  <span (click)=\"click()\" *ngIf=\"currentState !== states.Top && handle\"\n    [fivCenter]=\"{vertical: false, horizontal: true}\" class=\"handle\"></span>\n  <div [ngClass]=\"{'margin': handle}\" class=\"content\">\n    <ion-content [scrollY]=\"currentState === 2 && shouldBounce && !panning\">\n      <ng-content>\n\n      </ng-content>\n    </ion-content>\n\n  </div>\n</div>",
                styles: [":host{display:block;width:100%;height:100%;--fiv-border-radius:12px;--fiv-padding:20px;padding-top:var(--fiv-padding);--fiv-background:var(--ion-background-color);--fiv-handle-background:var(--ion-color-light-shade);background:var(--fiv-background)}.bsc-wrapper{width:100%;height:100%}:host.rounded{border-top-left-radius:var(--fiv-border-radius);border-top-right-radius:var(--fiv-border-radius)}.handle{display:block;background:var(--fiv-handle-background);border-radius:12px;top:8px;width:32px;height:8px;position:absolute}.content{height:100%}"]
            }] }
];
/** @nocollapse */
FivBottomSheetContent.ctorParameters = () => [
    { type: DomSanitizer }
];
FivBottomSheetContent.propDecorators = {
    rounded: [{ type: Input }],
    handle: [{ type: Input }],
    float: [{ type: Input }],
    fivClick: [{ type: Output }],
    content: [{ type: ViewChild, args: [IonContent, { static: true },] }],
    isRounded: [{ type: HostBinding, args: ['class.rounded',] }],
    styles: [{ type: HostBinding, args: ['style',] }]
};
if (false) {
    /** @type {?} */
    FivBottomSheetContent.prototype.rounded;
    /** @type {?} */
    FivBottomSheetContent.prototype.handle;
    /** @type {?} */
    FivBottomSheetContent.prototype.float;
    /** @type {?} */
    FivBottomSheetContent.prototype.fivClick;
    /** @type {?} */
    FivBottomSheetContent.prototype.content;
    /** @type {?} */
    FivBottomSheetContent.prototype.currentState;
    /** @type {?} */
    FivBottomSheetContent.prototype.states;
    /** @type {?} */
    FivBottomSheetContent.prototype.shouldBounce;
    /** @type {?} */
    FivBottomSheetContent.prototype.panning;
    /**
     * @type {?}
     * @private
     */
    FivBottomSheetContent.prototype.sanitizer;
}
