// @fivethree/core/esm2015/lib/stepper/step/step.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/stepper/step/step.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewChild, TemplateRef } from '@angular/core';
export class FivStep {
    constructor() {
        this.isLast = false;
        this.open = false;
        this.title = '';
        this.subtitle = '';
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
}
FivStep.decorators = [
    { type: Component, args: [{
                selector: 'fiv-step',
                template: "<ng-template>\n  <ng-content></ng-content>\n</ng-template>",
                styles: [""]
            }] }
];
/** @nocollapse */
FivStep.ctorParameters = () => [];
FivStep.propDecorators = {
    index: [{ type: Input }],
    icon: [{ type: Input }],
    isLast: [{ type: Input }],
    open: [{ type: Input }],
    title: [{ type: Input }],
    subtitle: [{ type: Input }],
    content: [{ type: ViewChild, args: [TemplateRef, { static: false },] }]
};
if (false) {
    /** @type {?} */
    FivStep.prototype.index;
    /** @type {?} */
    FivStep.prototype.icon;
    /** @type {?} */
    FivStep.prototype.isLast;
    /** @type {?} */
    FivStep.prototype.open;
    /** @type {?} */
    FivStep.prototype.title;
    /** @type {?} */
    FivStep.prototype.subtitle;
    /** @type {?} */
    FivStep.prototype.content;
}
