// @fivethree/core/esm5/lib/router-item/router-item.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/router-item/router-item.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { FivIcon } from './../icon/icon.component';
import { FivExpandable } from './../expandable/expandable.component';
import { Component, Input, HostBinding, HostListener, Optional, Host, ViewChild } from '@angular/core';
import { Router, NavigationEnd } from '@angular/router';
import { filter } from 'rxjs/operators';
var FivRouterItem = /** @class */ (function () {
    function FivRouterItem(router, host) {
        this.router = router;
        this.host = host;
        this.active = false;
        this.detailIcon = 'chevron-forward';
        this.disabled = false;
        this.matchChildUrl = false;
        this.navigate = true;
        /**
         * The position of the active state.
         * Default value is: `"left"`
         */
        this.position = 'left';
        /**
         * The shape of the active state.
         * Possible values are: `"line"`, `"dot"` and `"rounded"`.
         * Default value is: `"line"`
         */
        this.shape = 'line';
    }
    /**
     * @return {?}
     */
    FivRouterItem.prototype.onclick = /**
     * @return {?}
     */
    function () {
        if (this.navigate && this.pageUrl) {
            this.router.navigateByUrl(this.pageUrl);
        }
    };
    Object.defineProperty(FivRouterItem.prototype, "classes", {
        get: /**
         * @return {?}
         */
        function () {
            return this.position + " " + this.shape;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FivRouterItem.prototype, "activeClass", {
        get: /**
         * @return {?}
         */
        function () {
            return this.matchActiveUrl() || this.active;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FivRouterItem.prototype, "disabledClass", {
        get: /**
         * @return {?}
         */
        function () {
            return this.disabled;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    FivRouterItem.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @return {?}
     */
    FivRouterItem.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.router.events
            .pipe(filter((/**
         * @param {?} event
         * @return {?}
         */
        function (event) { return event instanceof NavigationEnd; })), filter((/**
         * @param {?} event
         * @return {?}
         */
        function (event) { return event.url === _this.pageUrl; })))
            .subscribe((/**
         * @return {?}
         */
        function () { return _this.openExpandableHost(); }));
    };
    /**
     * @return {?}
     */
    FivRouterItem.prototype.openExpandableHost = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var isActive = this.matchActiveUrl() || this.active;
        if (isActive) {
            if (this.host && !this.host.isOpen) {
                this.host.open();
            }
        }
    };
    /**
     * @return {?}
     */
    FivRouterItem.prototype.getClasses = /**
     * @return {?}
     */
    function () {
        return [this.position, this.shape];
    };
    /**
     * @return {?}
     */
    FivRouterItem.prototype.matchActiveUrl = /**
     * @return {?}
     */
    function () {
        return this.matchChildUrl ? this.isChildUrl() : this.isCurrentUrl();
    };
    /**
     * @return {?}
     */
    FivRouterItem.prototype.isCurrentUrl = /**
     * @return {?}
     */
    function () {
        return this.router.url === this.pageUrl;
    };
    /**
     * @return {?}
     */
    FivRouterItem.prototype.isChildUrl = /**
     * @return {?}
     */
    function () {
        return this.router.url.startsWith(this.pageUrl);
    };
    FivRouterItem.decorators = [
        { type: Component, args: [{
                    selector: 'fiv-router-item',
                    template: "<ion-menu-toggle auto-hide=\"false\">\n    <ion-item class=\"ion-activatable\" [color]=\"color\" [detail]=\"detail\" [detailIcon]=\"detailIcon\" [disabled]=\"disabled\"\n        [lines]=\"lines\">\n        <fiv-icon #fivIcon *ngIf=\"icon\" [name]=\"icon\"></fiv-icon>\n        <ion-label>{{text}}</ion-label>\n        <ion-ripple-effect type=\"bounded\"></ion-ripple-effect>\n    </ion-item>\n</ion-menu-toggle>\n<span [ngClass]=\"getClasses()\" class=\"indicator\"></span>",
                    styles: [":host{display:block;position:relative;width:100%}:host(.active){--color-active:var(--ion-color-primary)}:host(.active) ion-item{--color:var(--color-active);--background:var(--fiv-background, var(--ion-color-light))}:host(.line){--width-line:4px;--background-line:var(--ion-color-primary)}:host(.active) .line{display:block;background:var(--background-line)}:host.active.left .line{width:var(--width-line);height:100%;top:0}:host.active.right .line{width:var(--width-line);height:100%;top:0}.indicator{position:absolute}:host.active .dot{display:block}:host.active.dot.left{padding-left:24px}.dot{width:var(--fiv-dot-width,8px);height:var(--fiv-dot-heigt,8px);background-color:var(--fiv-dot-color,var(--ion-color-primary));top:50%;border-radius:100%;transform:translateY(-50%);right:8px;display:none}.dot.left{left:8px;right:unset}.dot.right{right:8px}:host(.rounded.left){--border-radius-rounded-left:0 50px 50px 0}:host(.rounded.left) ion-item{--border-radius:var(--border-radius-rounded-left);--background:var(--fiv-background)}:host(.rounded.right){--border-radius-rounded-right:50px 0 0 50px}:host(.rounded.right) ion-item{--border-radius:var(--border-radius-rounded-right);--background:var(--fiv-background)}:host(.active.rounded){--background-rounded:var(--ion-color-primary-rgb);--opacity-rounded:0.25}:host(.active.rounded) ion-item{--background:rgba(var(--background-rounded), var(--opacity-rounded))}@media (any-hover:hover){:host(:hover){cursor:pointer}:host(:hover) ion-label{--color:var(--ion-color-primary)}:host(.dot:hover) ion-item,:host(.line:hover) ion-item,:host(.rounded:hover) ion-item{--background:var(--fiv-background-hover, var(--ion-color-light))}}.left{left:0}.right{right:0}ion-item{--border-style:none}fiv-icon{--fiv-padding-end:8px}:host(.router-item-disabled){pointer-events:none}"]
                }] }
    ];
    /** @nocollapse */
    FivRouterItem.ctorParameters = function () { return [
        { type: Router },
        { type: FivExpandable, decorators: [{ type: Optional }, { type: Host }] }
    ]; };
    FivRouterItem.propDecorators = {
        fivIcon: [{ type: ViewChild, args: ['fivIcon', { static: false },] }],
        active: [{ type: Input }],
        color: [{ type: Input }],
        detail: [{ type: Input }],
        detailIcon: [{ type: Input }],
        disabled: [{ type: Input }],
        icon: [{ type: Input }],
        lines: [{ type: Input }],
        matchChildUrl: [{ type: Input }],
        navigate: [{ type: Input }],
        pageUrl: [{ type: Input }],
        position: [{ type: Input }],
        shape: [{ type: Input }],
        text: [{ type: Input }],
        onclick: [{ type: HostListener, args: ['click',] }],
        classes: [{ type: HostBinding, args: ['class',] }],
        activeClass: [{ type: HostBinding, args: ['class.active',] }],
        disabledClass: [{ type: HostBinding, args: ['class.router-item-disabled',] }]
    };
    return FivRouterItem;
}());
export { FivRouterItem };
if (false) {
    /** @type {?} */
    FivRouterItem.prototype.fivIcon;
    /** @type {?} */
    FivRouterItem.prototype.active;
    /** @type {?} */
    FivRouterItem.prototype.color;
    /** @type {?} */
    FivRouterItem.prototype.detail;
    /** @type {?} */
    FivRouterItem.prototype.detailIcon;
    /** @type {?} */
    FivRouterItem.prototype.disabled;
    /** @type {?} */
    FivRouterItem.prototype.icon;
    /** @type {?} */
    FivRouterItem.prototype.lines;
    /** @type {?} */
    FivRouterItem.prototype.matchChildUrl;
    /** @type {?} */
    FivRouterItem.prototype.navigate;
    /** @type {?} */
    FivRouterItem.prototype.pageUrl;
    /**
     * The position of the active state.
     * Default value is: `"left"`
     * @type {?}
     */
    FivRouterItem.prototype.position;
    /**
     * The shape of the active state.
     * Possible values are: `"line"`, `"dot"` and `"rounded"`.
     * Default value is: `"line"`
     * @type {?}
     */
    FivRouterItem.prototype.shape;
    /** @type {?} */
    FivRouterItem.prototype.text;
    /** @type {?} */
    FivRouterItem.prototype.router;
    /**
     * @type {?}
     * @private
     */
    FivRouterItem.prototype.host;
}
