// @fivethree/core/esm2015/lib/popover/popover-element.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/popover/popover-element.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { FivPopover } from './popover.component';
import { Directive, ElementRef, Input } from '@angular/core';
export class FivPopoverElementDirective {
    /**
     * @param {?} el
     */
    constructor(el) {
        this.el = el;
        this.ready = false;
        this.delay = 150;
    }
    /**
     * @param {?} visible
     * @return {?}
     */
    set show(visible) {
        this._show = visible;
        setTimeout((/**
         * @return {?}
         */
        () => {
            if (this.show && this.ready) {
                this.fivPopover.open(this.el);
            }
            else if (this.fivPopover.overlay && this.fivPopover.overlay.open) {
                this.fivPopover.close();
            }
        }), 0);
    }
    /**
     * @return {?}
     */
    get show() {
        return this._show;
    }
    /**
     * @return {?}
     */
    open() {
        this.show = true;
    }
    /**
     * @return {?}
     */
    close() {
        this.show = false;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        setTimeout((/**
         * @return {?}
         */
        () => {
            this.ready = true;
            this.show = this._show;
        }), this.delay);
    }
}
FivPopoverElementDirective.decorators = [
    { type: Directive, args: [{
                selector: '[fivPopover]',
                exportAs: 'fivPopover'
            },] }
];
/** @nocollapse */
FivPopoverElementDirective.ctorParameters = () => [
    { type: ElementRef }
];
FivPopoverElementDirective.propDecorators = {
    delay: [{ type: Input, args: ['fivPopover.delay',] }],
    fivPopover: [{ type: Input }],
    show: [{ type: Input, args: ['fivPopover.show',] }]
};
if (false) {
    /** @type {?} */
    FivPopoverElementDirective.prototype._show;
    /** @type {?} */
    FivPopoverElementDirective.prototype.ready;
    /** @type {?} */
    FivPopoverElementDirective.prototype.delay;
    /** @type {?} */
    FivPopoverElementDirective.prototype.fivPopover;
    /**
     * @type {?}
     * @private
     */
    FivPopoverElementDirective.prototype.el;
}
