// @fivethree/core/esm2015/lib/collapsable-menu/collapsable-menu-button/collapsable-menu-button.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/collapsable-menu/collapsable-menu-button/collapsable-menu-button.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { FivCollapsableMenu } from '../collapsable-menu.directive';
export class FivCollapsableMenuButton {
    constructor() {
        this.icon = 'menu';
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @return {?}
     */
    onClick() {
        this.menu.toggleMenu();
    }
}
FivCollapsableMenuButton.decorators = [
    { type: Component, args: [{
                selector: 'fiv-collapsable-menu-button',
                template: "<ion-button fivCenter (click)=\"onClick()\">\n  <fiv-icon slot=\"icon-only\" [name]=\"icon\" [color]=\"color\"></fiv-icon>\n</ion-button>",
                styles: [":host{display:block;padding-top:.2em}"]
            }] }
];
/** @nocollapse */
FivCollapsableMenuButton.ctorParameters = () => [];
FivCollapsableMenuButton.propDecorators = {
    menu: [{ type: Input }],
    icon: [{ type: Input }],
    color: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    FivCollapsableMenuButton.prototype.menu;
    /** @type {?} */
    FivCollapsableMenuButton.prototype.icon;
    /** @type {?} */
    FivCollapsableMenuButton.prototype.color;
}
